packages/itwin/tree-widget/src/components/trees/common/ReportingTreeEventHandler.ts

```diff
@@ -0,0 +1,89 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { concatMap, Observable as RxjsObservable, tap } from "rxjs";
+import { TreeEventHandler } from "@itwin/components-react";
+
+import type {
+  AbstractTreeNodeLoaderWithProvider,
+  TreeCheckboxStateChangeEventArgs,
+  TreeNodeEventArgs,
+  TreeSelectionChange,
+  TreeSelectionModificationEventArgs,
+  TreeSelectionReplacementEventArgs,
+} from "@itwin/components-react";
+import type { IPresentationTreeDataProvider } from "@itwin/presentation-components";
+export interface ReportingTreeEventHandlerProps {
+  nodeLoader: AbstractTreeNodeLoaderWithProvider<IPresentationTreeDataProvider>;
+  eventHandler?: TreeEventHandler;
+  reportUsage?: (props: { featureId?: string; reportInteraction: boolean }) => void;
+}
+
+export class ReportingTreeEventHandler extends TreeEventHandler {
+  protected _eventHandler?: TreeEventHandler;
+  protected _reportUsage?: (props: { featureId?: string; reportInteraction: boolean }) => void;
+
+  constructor(props: ReportingTreeEventHandlerProps) {
+    super({ modelSource: props.nodeLoader.modelSource, nodeLoader: props.nodeLoader });
+    this._eventHandler = props.eventHandler;
+    this._reportUsage = props.reportUsage;
+  }
+
+  public dispose() {
+    this._eventHandler?.dispose();
+    super.dispose();
+  }
+
+  public override onNodeExpanded(props: TreeNodeEventArgs) {
+    this._reportUsage?.({ reportInteraction: true });
+    this._eventHandler ? this._eventHandler.onNodeExpanded(props) : super.onNodeExpanded(props);
+  }
+
+  public override onNodeCollapsed(props: TreeNodeEventArgs) {
+    this._reportUsage?.({ reportInteraction: true });
+    this._eventHandler ? this._eventHandler.onNodeCollapsed(props) : super.onNodeCollapsed(props);
+  }
+
+  public override onSelectionModified(props: TreeSelectionModificationEventArgs) {
+    let emittedValues = false;
+    const rxjsObservable = new RxjsObservable<TreeSelectionChange>((subscriber) => props.modifications.subscribe(subscriber));
+    const tracked = rxjsObservable.pipe(
+      tap({
+        next: () => (emittedValues = true),
+        finalize: () => {
+          if (emittedValues) {
+            this._reportUsage?.({ reportInteraction: true });
+          }
+        },
+      }),
+      concatMap(() => rxjsObservable),
```

Comment:
Why do we need this special case only for `onSelectionModified`?

Reply:
Because for some reason the `onSelectionModified` is called when buttons in the hierarchy are clicked (expand node, visibility eye or filter buttons) but the passed observable does not emit any values, while when a node is selected it is called twice. I believe this also depends on the `SelectionMode` that is used. Added this to not report interaction multiple times.